### PR TITLE
Displaying Deterministic Username for a given Token

### PR DESCRIPTION
**Pull request by Kaz-ookid**
# PR Description

## Context
Throughout the app, users were previously referred to by their PoP Token, which was not user-friendly. This PR implements a deterministic username system to improve the user experience. A detailed documentation of the project's idea and motivations is to be found in #1915. It is recommended to read it before diving into this PR.

## Summary of Changes
This PR implements the deterministic username system across various parts of the app, ensuring a consistent and user-friendly display of usernames derived from PoP Tokens. The changes have been done in collaboration with FE2 (#1918) to ensure consistency throughout both systems.

### General Changes
- `PublicKey` class now includes the `getUsername()` function, returning the username version of the hash.
- Added a mnemonic deterministic username generator to `GeneralUtils`. The usernames are a concatenation of two mnemonics words, and a 4 digits number. The number corresponds to the 4 first digits in a token's hash, starting from the left

#### Example:
_token_ : `d_xeXEsurEnyWOp04mrrMxC3m4cS-3jK_9_Aw-UYfww=`
_mnemonics words generated form the hash_ : `spoon`, `issue`
_first 4 digits_ : `0434`
_final username_ : `spoonissue0434`

### Token
- **TokenFragment.kt**: When participating in a RollCall, you receive a PoP Token that you can see in the Token Tab. The username is now displayed above the hash of the token in the Token tab.

### Social Media
- **ChirpListAdapter.kt**: The chirps feed now displays the username of a user instead of the full PoP Token hash.

### Roll Call
- **RollCallArrayAdapter.kt**, **RollCallFragment.kt**: The list of attendees now displays the username above the Token hash. Structural changes were made in `RollCallArrayAdapter` to include two TextViews per item, and enhancing the class overall.

### Digital Cash
- Refactored the digital cash folder to work with `PublicKey` instead of `String` (token hashes).
- Displayed the username in appropriate places, simplifying the sending workflow and transaction history readability.
- For organizer transactions, the full public key hash is displayed with "(organizer)" appended, as the public key is longer than 32 bits and cannot be instantiated as `PublicKey`. Also, it juste made no sense to label this public key under a username anyway.
- Fixed a bug in the transaction history where title and value were inverted ("From" label, and Transaction ID). _see screenshots below_.
- Fixed a bug where the current user's token was not correctly removed from the selection Spinner when issuing as an organizer, causing a crash if it was selected.

### Bugs and Documentation
- Numerous bugs were discovered during implementation and have been well documented.

## Screenshots
- Token fragment displaying the username.
username_token

- Chirps feed displaying the username.
image

- Roll call displaying the username.
username_rollcall

- Digital cash sending activity displaying the username.
image

- Digital cash receipt activity displaying the username.
username_receipt

- Digital cash transaction history activity displaying the username and the organizer hash.
username_transactions

- Digital cash issuing activity.
image

- Bugged UI in transaction history showing the corrected title and value (now fixed).
image

### CODE REVIEW

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/security/PublicKey.kt [hunk 1]
```diff
@@ -14,13 +15,16 @@ import timber.log.Timber
 @Immutable
 class PublicKey : Base64URLData {
   private val verifier: PublicKeyVerify
+  private var toUsername: String
```
Comment (matteosz) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/security/PublicKey.kt:18]: Just a personal preference: maybe in the context of PublicKey I'd choose another name for that instead of username, such as keyLabel or simply label

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/lao/digitalcash/DigitalCashSendFragment.kt [hunk 2]
```diff
@@ -148,14 +151,20 @@ class DigitalCashSendFragment : Fragment() {
     binding.digitalCashSendSpinnerTv.setAdapter(adapter)
   }
 
+  private fun getPublicKeyFromUsername(username: String): PublicKey {
+    return digitalCashViewModel.attendeesFromTheRollCallList.firstOrNull {
+      it.getUsername() == username
+    } ?: digitalCashViewModel.validToken.publicKey
+  }
+
```
Comment (matteosz) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/lao/digitalcash/DigitalCashSendFragment.kt:159]: I see this function repeated across different files -> refactoring opportunity: create a static method that takes username, list of public key and alternative option if null. As the place I'd say either the PublicKey class itself or an utils class

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/security/PublicKey.kt (hunk 1 above)
Comment (matteosz) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/security/PublicKey.kt:18]: Also, do you want this field to be serialized? Because if not you should add a @Transient annotation

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/lao/digitalcash/HistoryListAdapter.kt [hunk 3]
```diff
@@ -51,7 +51,10 @@ class HistoryListAdapter(
     holder.transactionIdValue.text = transactionId
     holder.transactionProvenanceTitle.setText(
         if (element.isSender) R.string.digital_cash_to else R.string.digital_cash_from)
-    holder.transactionProvenanceValue.text = element.senderOrReceiver
+    holder.transactionProvenanceValue.text =
+        if (element.senderOrReceiver.encoded == viewModel.organizer.encoded)
+            element.senderOrReceiver.encoded + " (organizer)"
```
Comment (matteosz) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/lao/digitalcash/HistoryListAdapter.kt:56]: Shouldn't this be
`element.senderOrReceiver.getUsername()+ " (organizer)"`

File: fe2-android/app/src/test/framework/common/java/com/github/dedis/popstellar/testutils/pages/lao/digitalcash/HistoryPageObject.java [hunk 4]
```diff
@@ -14,4 +14,44 @@ private HistoryPageObject() {
...
+    //
+    //    // Check if the transaction details are displayed
+    //    onView(withId(HistoryPageObject.transactionProvenanceTitle())
+    //    )
+    //      .check(matches(isDisplayed()))
+    //    onView(withId(HistoryPageObject.transactionProvenanceValue())
+    //    )
+    //      .check(matches(isDisplayed()))
+    //    onView(withId(HistoryPageObject.transactionIdValue())
+    //    )
```
Comment (matteosz) [inline fe2-android/app/src/test/framework/common/java/com/github/dedis/popstellar/testutils/pages/lao/digitalcash/HistoryPageObject.java:31]: Do you need these comments to be here?

**Review by matteosz — CHANGES_REQUESTED** [review]
Great job! That's going to be a great feature :) 

Just some comments

**Comment by matteosz** [comment]
Also, now that I'm looking at the screenshots, perhaps it's better to have the username using underscores or capitalized letters so the 2 words are more easily recognizable
e.g. HomeSoldier0864, or home_soldier_0864, or any other fancy way of splitting the 2 words that you may prefer

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/security/PublicKey.kt (hunk 1 above)
Comment (Kaz-ookid) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/security/PublicKey.kt:18]: In fact the field should not be serialized. As it is deterministic, no need to send in anywhere since any system should be able to re-compute it in local themselves.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/lao/digitalcash/HistoryListAdapter.kt (hunk 3 above)
Comment (Kaz-ookid) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/lao/digitalcash/HistoryListAdapter.kt:56]: This is the case I explained in the PR description, where the Organizer key (used when Issuing) is longer than allowed for a `PublicKey`. Therefore we cannot use the `getUsername `function.

> For organizer transactions, the full public key hash is displayed with "(organizer)" appended, as the public key is longer than 32 bits and cannot be instantiated as PublicKey. Also, it juste made no sense to label this public key under a username anyway.

File: fe2-android/app/src/test/framework/common/java/com/github/dedis/popstellar/testutils/pages/lao/digitalcash/HistoryPageObject.java (hunk 4 above)
Comment (Kaz-ookid) [reply fe2-android/app/src/test/framework/common/java/com/github/dedis/popstellar/testutils/pages/lao/digitalcash/HistoryPageObject.java:31]: just some github copilot prompts I forgot to remove, nice catch!

**Comment by Kaz-ookid** [comment]
> Also, now that I'm looking at the screenshots, perhaps it's better to have the username using underscores or capitalized letters so the 2 words are more easily recognizable e.g. HomeSoldier0864, or home_soldier_0864, or any other fancy way of splitting the 2 words that you may prefer

I'll check with FE1 and modify before merging

**Review by matteosz — APPROVED** [review]

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/objects/security/KeyPairTest.kt [hunk 5]
```diff
@@ -50,6 +53,27 @@ class KeyPairTest {
...
+
+  @Test
+  fun pubKeyUsernameHashContainsLessDigits() {
+    val pk = PublicKey("oKHk3AivbpNXk_SfFcHDaVHcCcY8IBfHE7auXJmhmms=")
+    val digits = "0387"
+    // If the Hash contains less than 4 digits, the username will be padded with 0
+    Assert.assertEquals(digits, pk.getLabel().substring(pk.getLabel().length - USERNAME_DIGITS))
+  }
+
+  @Test
```
Comment (quadcopterman) [inline fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/objects/security/KeyPairTest.kt:72]: Maybe you could add one or two more tests then just the empty username for the generateUsername function

**Review by quadcopterman — COMMENTED** [review]
Looks good to me, just one comment

**Comment by quadcopterman** [comment]
Oh too late, it was merged while I was reviewing it
